### PR TITLE
KFP: add kubespin, kwasm support

### PR DESCRIPTION
**Pull request by dipankardas011**
# Tasks description

### Related Issue(s)
- #178 (KubeSpin support)

# Solution

### Sub-Tasks
- [ ] milestones or sub-task which needs to be coverted

# Note to reviewers

- [x] Ran Tests locally
- [x] Checked [Contribution's guidelines](https://docs.ksctl.com/docs/contribution-guidelines/)

### CODE REVIEW

**Comment by dipankardas011** [comment]
> [!NOTE]
> the version compatability is not checked and that is the responsibility of the user as of now

> [!IMPORTANT]
> possible solution is to use json mapping for component which is located in another repo which all of these call and get the metadata aka version compatability especially for production ready apps

**Comment by dipankardas011** [comment]
> Note
> 
> the version compatability is not checked and that is the responsibility of the user as of now
> 
> Important
> 
> possible solution is to use json mapping for component which is located in another repo which all of these call and get the metadata aka version compatability especially for production ready apps

for this we can refer for example **kwasm** <https://github.com/spinkube/containerd-shim-spin/pkgs/container/containerd-shim-spin%2Fnode-installer>

**Comment by dipankardas011** [comment]
we faced a blocker 🚫  (Helm package oci://)

references:
https://pkg.go.dev/helm.sh/helm/v3/pkg/registry#PullResult
https://pkg.go.dev/helm.sh/helm/v3/pkg/registry#PullOption
https://www.pulumi.com/registry/packages/kubernetes/api-docs/helm/v4/chart/
https://github.com/pulumi/pulumi-kubernetes/blob/434f2e8b8f5eb2144677aa65013f46a509b3cf0f/provider/pkg/helm/tool.go#L358

**Comment by dipankardas011** [comment]
Some similar examples found
Refs:
https://github.com/helm/helm/issues/12907
https://github.com/helm/helm-www/pull/1543/files

**Comment by dipankardas011** [comment]
Current Failure
```
wdfcsdf https://github.com/spinkube/spin-operator/releases/download/v0.2.0/spin-operator.shim-executor.yaml
{"time":"2024-08-10T06:00:59.628308667Z","level":"DEBUG","msg":"storage.kubernetes.Write","component":"store-kubernetes","Store":{}}
{"time":"2024-08-10T06:00:59.633154793Z","level":"DEBUG","msg":"configmap for write was found","component":"store-kubernetes"}
{"time":"2024-08-10T06:00:59.640363006Z","level":"ERROR","msg":"handled error","catch":"FailedKsctlComponentErr: failed to install component=kubernetes-client component=spinkube-operator-shim-executor Reason=FailedKubernetesClientErr: failed to decode the raw manifests into kubernetes gvr component=kubernetes-client Reason=no kind \"SpinAppExecutor\" is registered for version \"core.spinoperator.dev/v1alpha1\" in scheme \"pkg/runtime/scheme.go:100\""}
{"time":"2024-08-10T06:00:59.640425679Z","level":"INFO","msg":"K8s Storage Got Killed","component":"store-kubernetes","msgType":"SUCCESS"}
{"time":"2024-08-10T06:00:59.640438602Z","level":"ERROR","msg":"Handler","Reason":"FailedKsctlComponentErr: failed to install component=kubernetes-client component=spinkube-operator-shim-executor Reason=FailedKubernetesClientErr: failed to decode the raw manifests into kubernetes gvr component=kubernetes-client Reason=no kind \"SpinAppExecutor\" is registered for version \"core.spinoperator.dev/v1alpha1\" in scheme \"pkg/runtime/scheme.go:100\""}
```

**Comment by dipankardas011** [comment]
almost done just facing problems in these
```
{"time":"2024-08-10T13:37:24.550786446Z","level":"ERROR","msg":"handled error","catch":"FailedKsctlComponentErr: failed to install component=kubernetes-client component=spinkube-operator Reason=FailedHelmClientErr: failed to locate chart component=kubernetes-client Reason=repo spinkube not found"}
```

**Comment by dipankardas011** [comment]
left is running this
```shell
kubectl annotate node --all kwasm.sh/kwasm-node=true
```

**Comment by dipankardas011** [comment]
> ## Steps on How to run testing
> 
> Note
> 
> Given below steps help you to run the e2e tests on the PRs latest changes using the latest built images.
> 
> Warning
> 
> These images are built for testing purposes only and should not be used in production.
> 
> ```shell
> # first need to generate the application controller installer
> # make sure you don't push it to the PR
> make build-installer CONTROLLER="application" CUSTOM_LOCATION_GEN_CONTROLLER_MANIFEST="/tmp/ksctl-manifest.yaml" IMG_SUFFIX="/pr-429" IMG_TAG_VERSION="a10fc6a4ed9747c96ec9a82f4749ba7469f9d169"
> 
> cd test/e2e
> go build -v -ldflags="-X 'github.com/ksctl/ksctl/commons.OCIVersion=a10fc6a4ed9747c96ec9a82f4749ba7469f9d169' -X 'github.com/ksctl/ksctl/commons.OCIImgSuffix=/pr-429'" .
> export E2E_FLAGS="debug;core_component_overridings=application=file:::/tmp/ksctl-manifest.yaml" ## you can eliminate the debug
> ./e2e -op create -file local/create.json
> ./e2e -op switch -file local/switch.json
> ./e2e -op delete -file local/delete.json
> ```
> 
> Please make sure you've read our [contributing guide](CONTRIBUTING.md) and Happy Testing locally ✨

This works

**Comment by dipankardas011** [comment]
```yaml
apiVersion: application.ksctl.com/v1alpha1
kind: Stack
metadata:
  name: stack-sample-1
spec:
  stacks:
  - stackId: production-spinkube
    appType: app
```

**Comment by Horiodino** [comment]
> Should we give user the choice which all nodes to annotate?
> 
> ```shell
> kubectl annotate node --all kwasm.sh/kwasm-node=true
> ```
> 
> cc @saiyam1814 @Horiodino
can we just have no of nodes ,rather than this ? like no of nodes to use it for ? at the end the worker-node size is always same.

**Comment by Horiodino** [comment]
> Should we give user the choice which all nodes to annotate?
> 
> ```shell
> kubectl annotate node --all kwasm.sh/kwasm-node=true
> ```
> 
> cc @saiyam1814 @Horiodino

can we just have no of nodes ,rather than this ? like no of nodes to use it for ? at the end the worker-node size is always same.

**Comment by dipankardas011** [comment]
> ## Steps on How to run testing
> 
> Note
> 
> Given below steps help you to run the e2e tests on the PRs latest changes using the latest built images.
> 
> Warning
> 
> These images are built for testing purposes only and should not be used in production.
> 
> ```shell
> # first need to generate the application controller installer
> # make sure you don't push it to the PR
> make build-installer CONTROLLER="application" CUSTOM_LOCATION_GEN_CONTROLLER_MANIFEST="/tmp/ksctl-manifest.yaml" IMG_SUFFIX="/pr-429" IMG_TAG_VERSION="9516db352dc20310d7151e434e419e86f2a981d7"
> 
> cd test/e2e
> go build -v -ldflags="-X 'github.com/ksctl/ksctl/commons.OCIVersion=9516db352dc20310d7151e434e419e86f2a981d7' -X 'github.com/ksctl/ksctl/commons.OCIImgSuffix=/pr-429'" .
> export E2E_FLAGS="debug;core_component_overridings=application=file:::/tmp/ksctl-manifest.yaml" ## you can eliminate the debug
> ./e2e -op create -file local/create.json
> ./e2e -op switch -file local/switch.json
> ./e2e -op delete -file local/delete.json
> ```
> 
> Please make sure you've read our [contributing guide](CONTRIBUTING.md) and Happy Testing locally ✨

Updated now you can refer these steps

**Comment by dipankardas011** [comment]
> > Should we give user the choice which all nodes to annotate?
> > ```shell
> > kubectl annotate node --all kwasm.sh/kwasm-node=true
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > cc @saiyam1814 @Horiodino
> 
> can we just have no of nodes ,rather than this ? like no of nodes to use it for ? at the end the worker-node size is always same.

currently added all nodes annotation

**Comment by dipankardas011** [comment]
> can we just have no of nodes ,rather than this ? like no of nodes to use it for ? at the end the worker-node size is always same.

Opps missed your question,
we can think of this as well like some kind of `extraStackParameters`
Still not quite sure

**Comment by Horiodino** [comment]
> > can we just have no of nodes ,rather than this ? like no of nodes to use it for ? at the end the worker-node size is always same.
> 
> cc @Horiodino Opps missed your question, we can think of this as well like some kind of `extraStackParameters` Still not quite sure
> 
> Q) also another question should we have the stack contain single stack or multiple no of stacks?

should we give option .let the user  choose that   ?

**Comment by dipankardas011** [comment]
> should we give option .let the user choose that ?

what would be better is lets see that the community wants and if people are interested on these. for me what we can think of there are 3 types of overrides then one for component level specific ones, stack level overrides and finally once the postinstall overrides

### Scenarios
**Case 1**: it is a production stack
we will provide options for the components overrides but there will be somethings which stack will override for the component to work as a component for the specific job and if user tries to change these variables their change will be overriden. FInally for the postinstall overrides we can give users what happens after the install of stack like in this case nodes where we can install, labels to match,...

**Case 2**: it is development stack
as the stack in this context is very loosly related. Still it can contain more than one component For Example: argocd with argorollouts

> from all these we can note the boundary to make the app dev ready and prod ready is thin and we need to decide.

**Comment by dipankardas011** [comment]
### Another thing we need to consider

> How to get what is the latest version of a particular app.

**Possible Solutions**
1. Github api for releases
2. 3rd party or api driven method to get those

if its a option 1 we need to take into consideration about the rate limit.
for that I had a approach (Caching)we can have a go-routine which is kind of a subscribe the function to get the latest release based on cron schedule (may be 1h or 30m). it works like this if we try to get the `<org>/<repo>` it will first search in its `map[string][]string` if its present if not it will get the call and returns the res but at the same time stores in this memstore for it to poll on the next iteration of the poller based on `time.Ticker`

cc @Horiodino

**Comment by dipankardas011** [comment]
> ```shell
> make build-installer CONTROLLER="application" CUSTOM_LOCATION_GEN_CONTROLLER_MANIFEST="/tmp/ksctl-manifest.yaml" IMG_SUFFIX="/pr-429" IMG_TAG_VERSION="4de80a30a54dd77d7f253a76ba4841da474c53f7"
> 
> cd test/e2e
> go build -v -ldflags="-X 'github.com/ksctl/ksctl/commons.OCIVersion=4de80a30a54dd77d7f253a76ba4841da474c53f7' -X 'github.com/ksctl/ksctl/commons.OCIImgSuffix=/pr-429'" .
> export E2E_FLAGS="debug;core_component_overridings=application=file:::/tmp/ksctl-manifest.yaml" ## you can eliminate the debug
> ./e2e -op create -file local/create.json
> ./e2e -op switch -file local/switch.json
> ```

This Build is failing becuase None of the releases are stable 😕
image

Now we need to think of this if none of the release are stable we need to poll all the unstable releases

**Comment by dipankardas011** [comment]
> ## Steps on How to run testing
> 
> Note
> 
> Given below steps help you to run the e2e tests on the PRs latest changes using the latest built images.
> 
> Warning
> 
> These images are built for testing purposes only and should not be used in production.
> 
> ```shell
> # first need to generate the application controller installer
> # make sure you don't push it to the PR
> make build-installer CONTROLLER="application" CUSTOM_LOCATION_GEN_CONTROLLER_MANIFEST="/tmp/ksctl-manifest.yaml" IMG_SUFFIX="/pr-429" IMG_TAG_VERSION="b010305ae4f3af00ca61f2b09589816283a5f5a8"
> 
> cd test/e2e
> go build -v -ldflags="-X 'github.com/ksctl/ksctl/commons.OCIVersion=b010305ae4f3af00ca61f2b09589816283a5f5a8' -X 'github.com/ksctl/ksctl/commons.OCIImgSuffix=/pr-429'" .
> export E2E_FLAGS="debug;core_component_overridings=application=file:::/tmp/ksctl-manifest.yaml" ## you can eliminate the debug
> ./e2e -op create -file local/create.json
> ./e2e -op switch -file local/switch.json
> ./e2e -op delete -file local/delete.json
> ```
> 
> Please make sure you've read our [contributing guide](CONTRIBUTING.md) and Happy Testing locally ✨

Also Another edge case even if the published date can be newest its version can be old
for example: https://github.com/cert-manager/cert-manager/releases

**Comment by dipankardas011** [comment]
Done with kwasm and spin kube support

just need to create a stack with only kwasm

otherthan that it works now

**Comment by dipankardas011** [comment]
Here is an example for kwasm based wasmedge
```yaml
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: wasmedge
handler: wasmedge
---
apiVersion: v1
kind: Pod
metadata:
  name: "myapp"
  namespace: default
  labels:
    app: nice
spec:
  runtimeClassName: wasmedge
  containers:
  - name: myapp
    image: "docker.io/cr7258/wasm-demo-app:v1"
    ports:
    - containerPort: 8080
      name: http
  restartPolicy: Always
---
apiVersion: v1
kind: Service
metadata:
  name: nice
spec:
  selector:
    app: nice
  type: ClusterIP
  ports:
  - name: nice
    protocol: TCP
    port: 8080
    targetPort: 8080

```

**Comment by dipankardas011** [comment]
For this to work we need to create another stack for wasmedge & wasmtime(not quite sure) specifically

so currently when we installed spinkube it overrides default NodeInstaller with that of spin kube we need to have it custom the component from the stack itself, aka when it is getting called we will use this overrides to place it

File: internal/kubernetes/stacks/spinkube.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	if err != nil {
+		return metadata.ApplicationStack{}, err
+	}
+
+	// TODO: need to make delete work perfrrectly
+	//  Ctx: stkDepsIdx mode of traversal
+	//  Install:
+	//   - idx left to right
+	//  Uninstall:
+	//   - idx right to left
```
Comment (dipankardas011) [inline internal/kubernetes/stacks/spinkube.go:62]:
```suggestion
```

**Review by dipankardas011 — COMMENTED** [review]
Some changes otherthan that LGTM

**Review by dipankardas011 — COMMENTED** [review]
LGTM 🚀

**Pull request merged**